In [98]:
# Importing some libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np


In [99]:
#installing pandas_profiling for quick understanding of the data

In [5]:
pip install pandas-profiling 


Note: you may need to restart the kernel to use updated packages.


# Extracting the data

In [2]:
students=pd.read_csv("StudentsPerformance.csv")
students.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

# Checking the data information

In [3]:
students.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# Describing Data

In [4]:
students.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

In [100]:
from pandas_profiling import ProfileReport# import profile report

In [9]:
profile=ProfileReport(students,title="profiling report",explorative=True)

In [11]:
profile.to_widgets()

In [12]:
profile.to_file("your_report.html")


# Data Preparation

In [71]:
X = students.iloc[:, :-2].values
y = students.iloc[:, 5].values

In [72]:
from sklearn.model_selection import train_test_split

In [102]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)
#The above script splits 80% of the data to training set while 20% of the data to test set

# Training Algorithms

We have split our data into training and testing sets, and now is finally the time to train our algorithm. Execute following command:

In [74]:
from sklearn.linear_model import LinearRegression

In [75]:
reg=LinearRegression()

In [76]:
reg.fit(x_train,y_train)

LinearRegression()

In [77]:
print(reg.intercept_)

6.407508863812332


In [79]:
print(reg.coef_)

[0.86132413]


# Making Predictions

Now that we have trained our algorithm, it's time to make some predictions. To do so, we will use our test data and see how accurately our algorithm predicts the percentage score. To make pre-dictions on the test data, execute the following script:

In [80]:
y_predict=reg.predict(x_test)

In [83]:
y_predict

array([68.42284604, 69.28417017, 46.02841873, 59.80960477, 73.59079081,
       68.42284604, 65.83887366, 43.44444635, 81.34270795, 35.6925292 ,
       47.75106698, 66.70019779, 76.17476319, 75.31343906, 54.64166   ,
       41.72179809, 47.75106698, 84.78800446, 52.05768762, 73.59079081,
       62.39357715, 57.22563238, 71.86814255, 58.08695651, 46.02841873,
       62.39357715, 52.91901175, 44.30577047, 68.42284604, 83.92668034,
       81.34270795, 78.75873557, 86.51065272, 59.80960477, 55.50298413,
       74.45211493, 92.53992161, 56.36430826, 72.72946668, 68.42284604,
       61.53225302, 53.78033587, 70.1454943 , 73.59079081, 77.89741144,
       57.22563238, 84.78800446, 59.80960477, 77.03608732, 60.67092889,
       38.27650158, 59.80960477, 49.47371524, 54.64166   , 42.58312222,
       69.28417017, 52.05768762, 60.67092889, 71.00681842, 62.39357715,
       64.97754953, 74.45211493, 53.78033587, 72.72946668, 64.1162254 ,
       65.83887366, 69.28417017, 52.91901175, 66.70019779, 50.33

In [91]:
x_test.reshape(-1)

array([ 72,  73,  46,  62,  78,  72,  69,  43,  87,  34,  48,  70,  81,
        80,  56,  41,  48,  91,  53,  78,  65,  59,  76,  60,  46,  65,
        54,  44,  72,  90,  87,  84,  93,  62,  57,  79, 100,  58,  77,
        72,  64,  55,  74,  78,  83,  59,  91,  62,  82,  63,  37,  62,
        50,  56,  42,  73,  53,  63,  75,  65,  68,  79,  55,  77,  67,
        69,  73,  54,  70,  51,  71,  87,  73,  70,  56,  66,  70,  70,
        29,  79,  79,  89,  73,  85,  64,  77,  66,  87,  78,  62,  77,
        64,  65,  56,  76,  74,  48,  76,  66,  65,  51,  90,  64,  48,
        72,  80,  73,  65,  63,  91,  72,  75,  31,  59,  73,  76,  83,
        81,  82,  61,  56,  84,  93,  82,  74,  65,  74,  78,  68,  64,
        64,  64,  24,  81,  90,  58,  61,  64,  74,  82,  74,  68,  49,
        60,  86,  39,  43,  45,  84,  87,  49,  74,  99,  68,  82,  81,
        54,  58,  62,  52,  97,  59,  58,  93,  61,  83,  79,  66,  57,
        59,  61,  68,  39,  68,  83,  72,  68,  53, 100,  73,  6

In [92]:
y_predict.reshape(-1)

array([68.42284604, 69.28417017, 46.02841873, 59.80960477, 73.59079081,
       68.42284604, 65.83887366, 43.44444635, 81.34270795, 35.6925292 ,
       47.75106698, 66.70019779, 76.17476319, 75.31343906, 54.64166   ,
       41.72179809, 47.75106698, 84.78800446, 52.05768762, 73.59079081,
       62.39357715, 57.22563238, 71.86814255, 58.08695651, 46.02841873,
       62.39357715, 52.91901175, 44.30577047, 68.42284604, 83.92668034,
       81.34270795, 78.75873557, 86.51065272, 59.80960477, 55.50298413,
       74.45211493, 92.53992161, 56.36430826, 72.72946668, 68.42284604,
       61.53225302, 53.78033587, 70.1454943 , 73.59079081, 77.89741144,
       57.22563238, 84.78800446, 59.80960477, 77.03608732, 60.67092889,
       38.27650158, 59.80960477, 49.47371524, 54.64166   , 42.58312222,
       69.28417017, 52.05768762, 60.67092889, 71.00681842, 62.39357715,
       64.97754953, 74.45211493, 53.78033587, 72.72946668, 64.1162254 ,
       65.83887366, 69.28417017, 52.91901175, 66.70019779, 50.33

In [96]:
df=pd.DataFrame({'Actual':x_test.reshape(-1),'Predicted':y_predict.reshape(-1)})


In [97]:
df

Actual  Predicted
0        72  68.422846
1        73  69.284170
2        46  46.028419
3        62  59.809605
4        78  73.590791
..      ...        ...
195      61  58.948281
196      90  83.926680
197      83  77.897411
198      70  66.700198
199      67  64.116225

[200 rows x 2 columns]

# Evaluation of algorithms

In [101]:
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_predict.reshape(-1)))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_predict.reshape(-1)))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_predict.reshape(-1))))

Mean Absolute Error: 7.716733319559008
Mean Squared Error: 86.02373791316744
Root Mean Squared Error: 9.274898269693713


#You can see that the value of root mean squared error is 9.27, which is less than 15% of the mean value of the percentages of all the maths score i.e. 66.089. This means that our algorithm did a decent job.